velox/exec/tests/TableScanTest.cpp

```diff
@@ -1352,6 +1352,55 @@ TEST_F(TableScanTest, multipleSplits) {
   }
 }
 
+TEST_F(TableScanTest, preloadingSplitClose) {
+  auto filePaths = makeFilePaths(100);
+  auto vectors = makeVectors(100, 100);
+  for (int32_t i = 0; i < vectors.size(); i++) {
+    writeToFile(filePaths[i]->path, vectors[i]);
+  }
+  createDuckDbTable(vectors);
+
+  auto executors = ioExecutor_.get();
+  std::atomic<bool> stop = false;
+  std::atomic<int32_t> finishedCount = 0;
+  // Simulate a busy IO thread pool by blocking all threads.
+  for (int32_t i = 0; i < executors->numThreads(); ++i) {
+    executors->add([&stop, &finishedCount]() {
+      while (!stop) {
+        std::this_thread::sleep_for(std::chrono::milliseconds(50));
+      }
+      ++finishedCount;
+    });
+  }
+  ASSERT_EQ(Task::numCreatedTasks(), Task::numDeletedTasks());
+  auto task = assertQuery(tableScanNode(), filePaths, "SELECT * FROM tmp", 2);
+  auto stats = getTableScanRuntimeStats(task);
+
+  // Verify that split preloading is enabled.
+  ASSERT_GT(stats.at("preloadedSplits").sum, 1);
+
+  task.reset();
+  bool taskDelayedRelease = false;
+  // Even when the thread pool is blocked, tasks should be promptly destructed.
+  while (Task::numCreatedTasks() != Task::numDeletedTasks() &&
+         !taskDelayedRelease) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    taskDelayedRelease = true;
```

Comment:
I cannot figure out why `taskDelayedRelease` would be `false` below if we set it to `true` here.

Reply:
The expected behavior is that tasks are destructed immediately after they finish. So we should not execute the statements within the while loop.
I added this check to distinguish between the previous behavior: tasks being referenced by executor lambda and queued in the thread pool. Until the thread pool recovers, the lambdas are queued and tasks can't be released.
Thanks.

Reply:
I see. I am worried that thread timing issues will make this test flaky.
What if there is a slight delay in the task destruction and the while loop is executed?

Reply:
Task destructucton should be triggered by line 1382. as other references to task are removed by close() method I added in this PR. So I think there is no delay for destruction, as they are executed in the same thread. I added this check to ensure, task can be destructed after task finished, no matter whether the threadpool is busy or not.
Please let me know if I miss any thing here.
Thanks.

Reply:
> Why not simply use ASSERT_EQ(Task::numCreatedTasks(), Task::numDeletedTasks()); instead of this code block.

Make sense, updated, thanks.